rust/protocol/src/session_cipher.rs

```diff
@@ -238,11 +244,60 @@ pub async fn message_decrypt_signal<R: Rng + CryptoRng>(
     Ok(ptext)
 }
 
+fn create_decryption_failure_log(
+    remote_address: &ProtocolAddress,
+    errs: &[SignalProtocolError],
+    record: &SessionRecord,
+    ciphertext: &SignalMessage,
+) -> Result<String> {
+    let mut lines = vec![];
+
+    lines.push(format!(
+        "Message from {}:{} failed to decrypt; sender ratchet public key {} message counter {}",
+        remote_address.name(),
+        remote_address.device_id(),
+        hex::encode(ciphertext.sender_ratchet_key().public_key_bytes()?),
+        ciphertext.counter()
+    ));
+
+    for (idx, (state, err)) in std::iter::once(record.session_state()?)
+        .chain(record.previous_session_states()?)
+        .zip(errs)
+        .enumerate()
+    {
+        let chains = state.all_receiver_chain_logging_info()?;
+        lines.push(format!(
+            "Candidate session {} failed with '{}', had {} receiver chains",
+            idx,
+            err,
+            chains.len()
+        ));
+
+        for chain in chains {
+            let chain_idx = match chain.1 {
+                Some(i) => format!("{}", i),
+                None => "missing in protobuf".to_string(),
+            };
+
+            lines.push(format!(
+                "Receiver chain with sender ratchet public key {} chain key index {}",
```

Comment:
is the chain index in here the last counter received + 1 ? I believe that to be so but should be checked